gobra added

khoin opened the pull request:
In this PR:

I've made cobra to work with gobra.

I think you will need to `go get` skratchdot/open-golang first before it runs properly. I also don't think the binary will reflect this change unless you re-compile it, so on my machine, I tested it with `go run main.go` in /inmap.

It should look something like this: 

![image](https://user-images.githubusercontent.com/4620399/33395519-a1f2562a-d4fa-11e7-947c-7fa1498ffb90.png)

When the web interface is loaded, the model would run properly if you specify the config file. The output, is still dumped to the CLI:

![image](https://user-images.githubusercontent.com/4620399/33395593-e8253126-d4fa-11e7-9275-e42d006f9dca.png)

... But text generated by cobra would be dumped to the web interface: 
![image](https://user-images.githubusercontent.com/4620399/33395685-253195a0-d4fb-11e7-8682-df4b9792da7d.png)

About the code:

I saw that you intend to have the "HTTPPort" flag as a new flag for gobra. If you look into inmaputil/cmd.go at line 581, you'll see that I tried to not load the server if the HTTPPort flag is not specified. However, `Cfg.GetString("HTTPPort")` always return "8080" even when I run `go run main.go --HTTPPort=0`. I figured this is not important for now, so this PR goes through.

Let me know what you think.

khoin commented:
@ctessum

khoin commented:
Also, when you try to `run steady` on the web interface. You have to execute twice. I don't know why it is like that. I think when you run it the first time, it'll initialize the model, and the second time for the model to actually run.

ctessum commented:
Awesome!

One issue is that InMAP uses dependency vendoring, which causes some problems here at least for me when I try to run it. To fix this, install the [dep](https://github.com/golang/dep) program and then run `dep ensure` from the main inmap directory. This will make some changes to the `Gopkg.*` files, which you should check into git, and will also download a bunch of files to the 'vendor' directory, which you should not check in.

I'm not sure why it needs to be run twice to make it work, but I think it is important to fix it so it works the first time. Beyond confusing users, my concern is that even when it works sometimes it might be running with a configuration different from the one most recently specified.

ctessum commented on inmaputil/cmd.go:
```diff
@@ -569,6 +575,33 @@ func init() {
 	srCmd.AddCommand(srPredictCmd)
 	Root.AddCommand(workerCmd)
+
+	// For some reasons, HTTPPort flag never seem to be passed down?
```
This is because the flags are only evaluated when the cobra.command runs, not during global initialization. 

I think starting the GUI server during global initialization is causing the tests to fail. (You can see this in the test log output accompanying the pull request [the red stuff at the bottom of the conversation page) and also if you run `go test` in the inmaputil directory.

It also forces the user to use the GUI, they can no longer run InMAP from the command line. I think it would be better to start the server in `Root.Run()`. That way, the server only starts if the root command runs, which only happens if the user doesn't specify any subcommands at the command line. If the user specifies subcommands on the command line, then it would just run on the command line.

It would be a little tricky to do this, but I think it would work if you do it like this: https://play.golang.org/p/EEg5GAvQTv

khoin replied:
I actually initially started with gobra being inside Root.Run, but I ran into issues about circularity. I will think about this.

ctessum commented on inmaputil/cmd.go:
```diff
@@ -569,6 +575,33 @@ func init() {
+	if Cfg.IsSet("HTTPPort") && Cfg.GetString("HTTPPort") != "" { 
+		fmt.Println("Loading front-end...")
+
+		output := template.Must(template.New("").ParseFiles("../inmaputil/wrapper.html"))
```
When wrapper.html is a separate file, the program will only work when it is in the same directory as wrapper.html. Instead, we can include the wrapper html as a text string within the go source code, e.g., 
``` go
const wrapper = `doctype html etc...`
```

khoin replied:
Alright, I will do that. This will assume that you are OK with the long wrapper.html being in the source code.

ctessum commented on inmaputil/cmd.go:
```diff
@@ -569,6 +575,33 @@ func init() {
+		output.ExecuteTemplate(f, "base", template.HTML(gobraInterface))
+
+		server := gobra.Server{Root, Cfg.GetInt("HTTPPort"), false, false}
+		fmt.Println("Server starting... ")
```
Probably better to use `log` instead of `fmt` so we get a timestamp

ctessum commented on inmaputil/wrapper.html:
```diff
@@ -0,0 +1,27 @@
+{{ define "base" }}
```
As above, include all of this as a string constant in the go source code

ctessum commented on inmaputil/wrapper.html:
```diff
@@ -0,0 +1,27 @@
+<html>
+<head>
+	<meta charset="utf-8">
+	<title>InMap Web Interface</title>
```
Maybe just InMAP, so people don't think this is coming to them over then internet?

khoin commented on inmap/index.html:
```diff
@@ -0,0 +1,804 @@
+
+<!DOCTYPE html>
```
I was bothered by this as well. I will look into that. Is this PR enough for your presentation on Monday?

ctessum replied:
Hmm. How about if you're able to make all the superficial changes here, I can deal with the rest of them

khoin replied:
If I get the room key from CE frontdesk on Friday, ETA for all of these should be Friday afternoon (your Saturday), except for https://github.com/ctessum/gobra/issues/11 and https://github.com/spatialmodel/inmap/pull/30#discussion_r153989808 . These two will take probably take more time.

ctessum commented on inmaputil/wrapper.html:
```diff
@@ -0,0 +1,27 @@
+	<style>
+		html, body {padding: 0; margin: 2% 0; font-family: sans-serif;}
+		.container { max-width: 700px; margin: 0 auto; padding: 10px; }
+		blockquote { border-left: 5px solid #aaa; margin: .3em; color: #333; padding: 3px; }
```
Can we make the text smaller for the block quote and use a fixed-width font for the editable fields, and maybe bold text for the flag names, with the goal being to make it more clear which information is most important and where data can be entered?

ctessum commented on inmaputil/wrapper.html:
```diff
@@ -0,0 +1,27 @@
+</head>
+<body>
+<div class="container">
+	<h1>InMap Web Interface</h1>
```
Maybe just "InMAP" again

ctessum commented on inmaputil/wrapper.html:
```diff
@@ -0,0 +1,27 @@
+<body>
+<div class="container">
+	<h1>InMap Web Interface</h1>
+	<span>Navigate your commands below</span>
```
Maybe "Configure the simulation below:"

ctessum commented on inmaputil/wrapper.html:
```diff
@@ -0,0 +1,27 @@
+		{{.}}
+	</div>
+	<footer>
+		Footer Text.
```
© 2017 InMAP Authors

ctessum commented:
Also, can you add your name to CONTRIBUTORS.md and add an entry to CHANGELOG.md?

khoin commented on inmaputil/cmd.go:
```diff
@@ -560,6 +563,9 @@ func init() {
 // Link the commands together.
+// Also generate Gobra interface
+
+var gobraInterface string;
```
Rookie mistake. Will be fixed.

ctessum commented:
Great. Let me know if there is anything I can do to expedite the room key
issue.

On Thu, Nov 30, 2017, 2:32 PM khoin <notifications@github.com> wrote:

> *@khoin* commented on this pull request.
> ------------------------------
>
> In inmap/index.html
> <https://github.com/spatialmodel/inmap/pull/30#discussion_r153995033>:
>
> > @@ -0,0 +1,804 @@
> +
> +<!DOCTYPE html>
>
> If I get the room key from CE frontdesk on Friday, ETA for all of these
> should be Friday afternoon (your Saturday), except for ctessum/gobra#11
> <https://github.com/ctessum/gobra/issues/11> and #30 (comment)
> <https://github.com/spatialmodel/inmap/pull/30#discussion_r153989808> .
> These two will take probably take more time.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/spatialmodel/inmap/pull/30#discussion_r153995033>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AF6AAU2qY24D5L_N59pwC5PkuJU4vwBkks5s7kv7gaJpZM4Qvh4a>
> .
>

khoin commented:
I made some changes:

* Merged your PR on my fork. (https://github.com/khoin/inmap/pull/1)
* Bring in dep. Also, check out the things I mentioned in your PR. I think I'm having problems with dep
* Superficial changes for wrapper.html
* Deleted wrapper.html and moved it into the source code.
* CONTRIBUTORS & Changelog

khoin commented:
I've merged the PR on gobra. I've also merged your PR on my inmap fork. I've squashed all the commits into one now.

Most issues mentioned in this PR should be resolved. Other issues more related to Gobra will be resolved in that repo.

Your turn.

ctessum commented:
Great. Thanks!
